src/System.IO.Pipes/src/System/IO/Pipes/NamedPipeClientStream.Windows.cs

```diff
@@ -79,14 +56,58 @@ private bool TryConnect(int timeout, CancellationToken cancellationToken)
             {
                 int errorCode = Marshal.GetLastWin32Error();
 
-                // Handle the possible race condition of someone else connecting to the server 
-                // between our calls to WaitNamedPipe & CreateFile.
-                if (errorCode == Interop.Errors.ERROR_PIPE_BUSY)
+                if (errorCode != Interop.Errors.ERROR_PIPE_BUSY &&
+                    errorCode != Interop.Errors.ERROR_FILE_NOT_FOUND)
+                {
+                    throw Win32Marshal.GetExceptionForWin32Error(errorCode);
+                }
+
+                if (!Interop.Kernel32.WaitNamedPipe(_normalizedPipePath, timeout))
                 {
-                    return false;
+                    errorCode = Marshal.GetLastWin32Error();
+
+                    // Server is not yet created
+                    if (errorCode == Interop.Errors.ERROR_FILE_NOT_FOUND)
+                    {
+                        return false;
+                    }
+
+                    // The timeout has expired.
+                    if (errorCode == Interop.Errors.ERROR_SUCCESS)
+                    {
+                        if (cancellationToken.CanBeCanceled)
+                        {
+                            // It may not be real timeout.
+                            return false;
+                        }
+                        throw new TimeoutException();
+                    }
+
+                    throw Win32Marshal.GetExceptionForWin32Error(errorCode);
                 }
 
-                throw Win32Marshal.GetExceptionForWin32Error(errorCode);
+                // Pipe server should be free.  Let's try to connect to it.
+                handle = Interop.Kernel32.CreateNamedPipeClient(_normalizedPipePath,
```

Comment:
The reason I wanted you to add a second call here is because it was possible that the outer loop will bail if it crosses the elapsed time, which we would possibly allow in the prior code. Pretty slim chance, but still...